lib/tenstorrent/bh_chip/bh_chip.c

```diff
@@ -145,9 +147,82 @@ int therm_trip_gpio_setup(struct bh_chip *chip)
 	if (ret != 0) {
 		return ret;
 	}
-	gpio_init_callback(&chip->therm_trip_cb, therm_trip_detected,
+	gpio_init_callback(&chip->data.therm_trip_cb, therm_trip_detected,
 			   BIT(chip->config.therm_trip.pin));
-	ret = gpio_add_callback(chip->config.therm_trip.port, &chip->therm_trip_cb);
+	ret = gpio_add_callback_dt(&chip->config.therm_trip, &chip->data.therm_trip_cb);
+
+	return ret;
+}
+
+void pgood_fault_work_handler(struct k_work *work)
+{
+	struct bh_chip_data *data = CONTAINER_OF(work, struct bh_chip_data, pgood_fault_worker);
+	struct bh_chip *chip = CONTAINER_OF(data, struct bh_chip, data);
+	static const struct gpio_dt_spec board_fault_led =
+		GPIO_DT_SPEC_GET_OR(DT_PATH(board_fault_led), gpios, {0});
+
+	k_sem_reset(&chip->data.pgood_high_sem);
```

Comment:
What about doing this in the callback:

* try to take a semaphore with `K_NO_WAIT`
* if the return value is 0 then all is good
* otherwise if it's `EAGAIN`, just return - it means a reset is already in progress
* schedule delayed work for 50ms
* in the delayed work handler, give the semaphore and do the pgood check below where you assert the board fault, etc